File: stackunderflow/adapters/pi.py
"""Pi (and OMP) session adapter.

Pi and OMP are sibling CLIs that share an on-disk format but use
different roots:

- Pi: ``~/.pi/agent/sessions/``
- OMP: ``~/.omp/agent/sessions/``

We implement them as a **single** ``PiAdapter`` that scans both roots —
the diff between two adapters would be one constant — and the env-flag
``STACKUNDERFLOW_BETA_PI`` controls discovery for both. ``project_slug``
embeds the source root name (``pi`` or ``omp``) so cross-tool reports
can still split them apart.

JSONL events look like::

    {"type": "session", "id": "...", "timestamp": "...", "cwd": "..."}
    {"type": "message", "id": "...", "timestamp": "...",
     "message": {"role": "assistant",
                 "content": [{"type": "text", "text": "..."}],
                 "model": "gpt-5",
                 "responseId": "...",
                 "usage": {"input": ..., "output": ...,
                           "cacheRead": ..., "cacheWrite": ...}}}

Storage: byte-offset resume (spec §1.4) — same as Codex.

Spec §3 (multi-provider).

Defensive sizing: JSONL sessions larger than ``MAX_SESSION_FILE_BYTES``
(128 MB; see ``stackunderflow/adapters/_streaming.py``) are **skipped
with a logged warning** rather than parsed. Smaller files stream
line-by-line."""

from __future__ import annotations

import json
import logging
from collections.abc import Iterator
from pathlib import Path

from ._streaming import iter_jsonl_lines
from .base import Record, SessionRef

_log = logging.getLogger(__name__)

_DEFAULT_MODEL = "gpt-5"

# (root path, label) — label feeds project_slug so a downstream report
# can keep Pi and OMP distinct without re-deriving from the path.
_DEFAULT_ROOTS: tuple[tuple[Path, str], ...] = (
    (Path.home() / ".pi" / "agent" / "sessions", "pi"),
    (Path.home() / ".omp" / "agent" / "sessions", "omp"),
)


class PiAdapter:
    """Source adapter for Pi and OMP CLIs (shared format)."""

    name = "pi"

    def __init__(self, roots: list[tuple[Path, str]] | None = None) -> None:
        # Tests inject explicit roots; production scans both defaults.
        self._roots = list(roots) if roots is not None else list(_DEFAULT_ROOTS)

    # ── enumeration ───────────────────────────────────────────────────

    def enumerate(self) -> Iterator[SessionRef]:
        for root, label in self._roots:
            if not root.is_dir():
                continue
            for fp in sorted(root.glob("**/*.jsonl")):
                try:
                    stat = fp.stat()
                except OSError as exc:
                    _log.warning("Cannot stat Pi/OMP session %s: %s", fp, exc)
                    continue

                session_id, cwd = _peek_session_meta(fp)
                if not session_id:
                    session_id = fp.stem
                project_slug = _slug_for(cwd, label) if cwd else label

                yield SessionRef(
                    provider=self.name,
                    project_slug=project_slug,
                    session_id=session_id,
                    file_path=fp,
                    file_mtime=stat.st_mtime,
                    file_size=stat.st_size,
                    source_kind="file",
                    # Hint preserves the source label so downstream code
                    # can tell Pi sessions from OMP sessions without
                    # re-parsing the file path.
                    source_hint={"source": label},
                )

    def watch_paths(self) -> list[Path]:
        return [root for root, _ in self._roots]

    # ── reading ───────────────────────────────────────────────────────

    def read(self, ref: SessionRef, *, since_offset: int = 0) -> Iterator[Record]:
        # ``iter_jsonl_lines`` enforces the 128 MB defensive cap and
        # streams line-by-line; oversize sessions are skipped with a
        # warning rather than parsed.
        for line_offset, raw_line in iter_jsonl_lines(
            ref.file_path, since_offset=since_offset,
        ):
            if since_offset > 0 and line_offset <= since_offset:
                continue
            stripped = raw_line.strip()
            if not stripped:
                continue
            try:
                event = json.loads(stripped)
            except (json.JSONDecodeError, ValueError) as exc:
                _log.debug(
                    "Skipping malformed JSON line in %s: %s",
                    ref.file_path, exc,
                )
                continue
            if not isinstance(event, dict):
                # Valid JSON that isn't an object (list / string / number)
                # can't be a session event — skip, don't crash the read.
                continue

            if event.get("type") != "message":
                continue

            message = event.get("message") or {}
            if not isinstance(message, dict):
                continue
            if message.get("role") != "assistant":
                continue
            usage = message.get("usage")
            if not isinstance(usage, dict):
                continue

            model = (
                str(message.get("model"))
                if isinstance(message.get("model"), str)
                and message.get("model")
                else _DEFAULT_MODEL
            )

            tokens = _normalize_usage(usage)
            content = message.get("content")
            cwd = event.get("cwd")

            yield Record(
                provider=self.name,
                session_id=ref.session_id,
                seq=line_offset,
                timestamp=str(event.get("timestamp") or ""),
                role="assistant",
                model=model,
                input_tokens=tokens["input"],
                output_tokens=tokens["output"],
                cache_create_tokens=tokens["cache_creation"],
                cache_read_tokens=tokens["cache_read"],
                content_text=_message_text(content),
                tools=_tools_from_content(content),
                cwd=cwd if isinstance(cwd, str) and cwd else None,
                is_sidechain=False,
                uuid=str(event.get("id") or f"{ref.session_id}:{line_offset}"),
                parent_uuid=None,
                raw=event,
            )


# ── helpers ───────────────────────────────────────────────────────────


def _peek_session_meta(fp: Path) -> tuple[str, str]:
    """Return ``(session_id, cwd)`` from the first ``session`` event."""
    try:
        with fp.open("rb") as fh:
            first = fh.readline()
    except OSError:
        return "", ""
    stripped = first.strip()
    if not stripped:
        return "", ""
    try:
        obj = json.loads(stripped)
    except (json.JSONDecodeError, ValueError):
        return "", ""
    if not isinstance(obj, dict):
        # A non-object first line must not crash enumerate() — fall back
        # to the filename-stem session id.
        return "", ""
    if obj.get("type") != "session":
        return "", ""
    return str(obj.get("id") or ""), str(obj.get("cwd") or "")


def _slug_for(project_path: str, label: str) -> str:
    """Project slug includes the source label so Pi vs OMP stays separate."""
    import os

    cleaned = (
        os.path.abspath(project_path)
        .rstrip(os.sep)
        .replace(os.sep, "-")
        .replace("_", "-")
    )
    return f"{label}{cleaned}"


def _normalize_usage(usage: dict) -> dict[str, int]:
    """Pi/OMP shape → canonical 4-key shape."""
    return {
        "input": _safe_int(usage.get("input")),
        "output": _safe_int(usage.get("output")),
        "cache_creation": _safe_int(usage.get("cacheWrite")),
        "cache_read": _safe_int(usage.get("cacheRead")),
    }


def _safe_int(val: object) -> int:
    if isinstance(val, (int, float)):
        try:
            return max(int(val), 0)
        except (OverflowError, ValueError):
            # float('inf') / float('nan') — JSON like ``1e999`` parses to
            # inf; int() on it raises instead of coercing.
            return 0
    if isinstance(val, str):
        try:
            return max(int(val), 0)
        except ValueError:
            return 0
    if isinstance(val, bytes | bytearray):
        try:
            return max(int(val.decode("utf-8", errors="replace")), 0)
        except ValueError:
            return 0
    return 0


def _message_text(content: object) -> str:
    if isinstance(content, str):
        return content
    if not isinstance(content, list):
        return ""
    pieces: list[str] = []
    for blk in content:
        if isinstance(blk, dict):
            t = blk.get("text")
            if isinstance(t, str) and t:
                pieces.append(t)
        elif isinstance(blk, str):
            pieces.append(blk)
    return "\n".join(pieces)


def _tools_from_content(content: object) -> tuple[str, ...]:
    """Pluck tool names from ``toolCall`` / ``tool_use`` blocks."""
    if not isinstance(content, list):
        return ()
    tools: list[str] = []
    for blk in content:
        if not isinstance(blk, dict):
            continue
        if blk.get("type") in ("toolCall", "tool_use"):
            name = blk.get("name")
            if isinstance(name, str) and name:
                tools.append(name)
    return tuple(tools)

File: rust/crates/stax-adapters/src/pi.rs
//! Pi and OMP — the port of `python-legacy: adapters/pi.py`.
//!
//! Two sibling CLIs with one on-disk format and two roots:
//!
//! * Pi — `~/.pi/agent/sessions/`
//! * OMP — `~/.omp/agent/sessions/`
//!
//! They are **one adapter**, because the difference between two would be a
//! single constant. What keeps them distinguishable downstream is the
//! `project_slug`, which carries the source label as a prefix
//! (`pi-Users-me-app`, `omp-Users-me-app`), plus a `source_hint` of
//! `{"source": "pi"|"omp"}` so a report never has to re-derive it from the file
//! path.
//!
//! Events are one JSON object per line:
//!
//! ```text
//! {"type": "session", "id": "...", "timestamp": "...", "cwd": "..."}
//! {"type": "message", "id": "...", "timestamp": "...",
//!  "message": {"role": "assistant", "content": [...], "model": "gpt-5",
//!              "usage": {"input": …, "output": …, "cacheRead": …, "cacheWrite": …}}}
//! ```
//!
//! Only assistant messages **that carry a usage block** become records —
//! everything else, user turns included, yields nothing. `seq` is the byte
//! offset of the line, so resume is the same seek-and-compare as every other
//! JSONL adapter.

use std::path::{Path, PathBuf};

use serde_json::{Map, Value};

use crate::base::{Record, SessionRef, SourceAdapter, Speed, stat_ref_fields};
use crate::jsonl::{JsonlLines, parse_json, py_bytes_strip};
use crate::{blocks, pyval, walk};

/// The provider key. One name for both CLIs — see the module docs.
pub const NAME: &str = "pi";

/// The model stamped on a turn whose `message.model` is missing or non-string
/// (`_DEFAULT_MODEL`).
///
/// Not `None`: a model-less record is dropped by the normalizer as unpriceable,
/// and Pi's own default really is this model.
pub const DEFAULT_MODEL: &str = "gpt-5";

/// The block types that count as a tool call (`_tools_from_content`).
pub const TOOL_BLOCK_TYPES: [&str; 2] = ["toolCall", "tool_use"];

/// The Pi / OMP source adapter (`PiAdapter`).
#[derive(Debug, Clone)]
pub struct PiAdapter {
    /// `(root, label)` pairs — the label feeds `project_slug` and `source_hint`.
    roots: Vec<(PathBuf, String)>,
}

impl Default for PiAdapter {
    fn default() -> Self {
        Self::new()
    }
}

impl PiAdapter {
    /// Both default roots, resolved against the home directory.
    ///
    /// Python evaluates `_DEFAULT_ROOTS` once at import; resolving once at
    /// construction is the same observable behaviour without a global.
    #[must_use]
    pub fn new() -> Self {
        #[allow(
            deprecated,
            reason = "std::env::home_dir is the platform-correct answer on the \
            1.97.1 pin; the claude and codex adapters carry the same allow"
        )]
        let home = std::env::home_dir().unwrap_or_default();
        Self {
            roots: vec![
                (home.join(".pi").join("agent").join("sessions"), "pi".into()),
                (
                    home.join(".omp").join("agent").join("sessions"),
                    "omp".into(),
                ),
            ],
        }
    }

    /// Inject explicit `(root, label)` pairs — the constructor parameter Python
    /// already has (`PiAdapter(roots=[(tmp, "pi")])`).
    #[must_use]
    pub fn with_roots(roots: Vec<(PathBuf, String)>) -> Self {
        Self { roots }
    }

    /// The `(root, label)` pairs this adapter scans.
    #[must_use]
    pub fn roots(&self) -> &[(PathBuf, String)] {
        &self.roots
    }
}

impl SourceAdapter for PiAdapter {
    fn name(&self) -> &str {
        NAME
    }

    fn enumerate(&self) -> Vec<SessionRef> {
        let cwd = current_dir_string();
        let mut out = Vec::new();
        for (root, label) in &self.roots {
            if !root.is_dir() {
                continue;
            }
            // `sorted(root.glob("**/*.jsonl"))` — this directory and every
            // subdirectory, sorted by path string.
            for path in walk::rglob_suffix(root, ".jsonl") {
                // Python warns and continues on a stat failure; a session that
                // vanished mid-walk must not take the other roots down.
                let Some((mtime, size)) = stat_ref_fields(&path) else {
                    continue;
                };
                let (peeked_id, peeked_cwd) = peek_session_meta(&path);
                let session_id = if peeked_id.is_empty() {
                    walk::file_stem(&path)
                } else {
                    peeked_id
                };
                let project_slug = if peeked_cwd.is_empty() {
                    label.clone()
                } else {
                    slug_for(&peeked_cwd, label, &cwd)
                };
                let mut hint = Map::new();
                hint.insert("source".to_string(), Value::from(label.clone()));
                out.push(SessionRef {
                    provider: NAME.to_string(),
                    project_slug,
                    session_id,
                    file_path: path,
                    file_mtime: mtime,
                    file_size: size,
                    source_kind: crate::base::SourceKind::File,
                    source_hint: Some(hint),
                });
            }
        }
        out
    }

    fn read_into(&self, session: &SessionRef, since_offset: i64, sink: &mut dyn FnMut(Record)) {
        // No up-front size check: `JsonlLines` enforces the 128 MB cap and
        // yields nothing above it, which is exactly what `iter_jsonl_lines`
        // does for the Python original.
        for (line_offset, raw_line) in JsonlLines::open(&session.file_path, since_offset) {
            if since_offset > 0 && line_offset <= since_offset {
                continue;
            }
            let stripped = py_bytes_strip(&raw_line);
            if stripped.is_empty() {
                continue;
            }
            // LOG: python debug-logs "Skipping malformed JSON line in %s".
            let Some(event) = parse_json(stripped) else {
                continue;
            };
            // Valid JSON that is not an object cannot be a session event.
            let Some(map) = event.as_object() else {
                continue;
            };
            if map.get("type").and_then(Value::as_str) != Some("message") {
                continue;
            }
            // `event.get("message") or {}` then an isinstance check: a falsy
            // message (null, "", []) becomes `{}`, whose role is absent, so
            // both paths land on "skip".
            let Some(message) = map.get("message").and_then(Value::as_object) else {
                continue;
            };
            if message.get("role").and_then(Value::as_str) != Some("assistant") {
                continue;
            }
            // A turn without a usage *dict* is not billable and yields nothing
            // at all — not a zero-token record.
            let Some(usage) = message.get("usage").and_then(Value::as_object) else {
                continue;
            };

            let model = message
                .get("model")
                .and_then(Value::as_str)
                .filter(|value| !value.is_empty())
                .map_or_else(|| DEFAULT_MODEL.to_string(), ToString::to_string);
            let content = message.get("content");
            let uuid = map
                .get("id")
                .filter(|value| pyval::py_truthy(value))
                .map_or_else(
                    || format!("{}:{line_offset}", session.session_id),
                    pyval::py_str,
                );

            sink(Record {
                provider: NAME.to_string(),
                session_id: session.session_id.clone(),
                seq: line_offset,
                timestamp: map
                    .get("timestamp")
                    .filter(|value| pyval::py_truthy(value))
                    .map_or_else(String::new, pyval::py_str),
                role: "assistant".to_string(),
                model: Some(model),
                input_tokens: pyval::safe_int(usage.get("input")),
                output_tokens: pyval::safe_int(usage.get("output")),
                cache_create_tokens: pyval::safe_int(usage.get("cacheWrite")),
                cache_read_tokens: pyval::safe_int(usage.get("cacheRead")),
                content_text: blocks::message_text(content),
                tools: blocks::tool_names(content, &TOOL_BLOCK_TYPES),
                // The cwd lives on the *event*, not on the message.
                cwd: map
                    .get("cwd")
                    .and_then(Value::as_str)
                    .filter(|value| !value.is_empty())
                    .map(ToString::to_string),
                is_sidechain: false,
                uuid,
                parent_uuid: None,
                raw: event.clone(),
                speed: Speed::Standard,
            });
        }
    }

    fn watch_paths(&self) -> Vec<PathBuf> {
        // Both roots unconditionally — the watcher filters non-existent ones,
        // so a machine with only Pi installed still picks OMP up the day it is.
        self.roots.iter().map(|(root, _)| root.clone()).collect()
    }
}

/// `(session_id, cwd)` from the file's first line (`_peek_session_meta`).
///
/// Empty strings for every failure: unreadable file, blank first line, invalid
/// JSON, a first line that is not an object, or a first event that is not the
/// `session` header. The caller falls back to the filename stem.
fn peek_session_meta(path: &Path) -> (String, String) {
    let empty = || (String::new(), String::new());
    let Some(first) = walk::first_line(path) else {
        return empty();
    };
    let stripped = py_bytes_strip(&first);
    if stripped.is_empty() {
        return empty();
    }
    let Some(obj) = parse_json(stripped) else {
        return empty();
    };
    let Some(map) = obj.as_object() else {
        return empty();
    };
    if map.get("type").and_then(Value::as_str) != Some("session") {
        return empty();
    }
    let field = |key: &str| {
        map.get(key)
            .filter(|value| pyval::py_truthy(value))
            .map_or_else(String::new, pyval::py_str)
    };
    (field("id"), field("cwd"))
}

/// `<label><claude-style slug>` (`_slug_for`).
///
/// The label prefix is what keeps a Pi session and an OMP session in the same
/// directory from collapsing into one project.
fn slug_for(project_path: &str, label: &str, cwd: &str) -> String {
    format!("{label}{}", pyval::slug_for(project_path, cwd))
}

/// `os.getcwd()` for the slug derivation; `"/"` when the process has no cwd.
fn current_dir_string() -> String {
    std::env::current_dir().map_or_else(
        |_| "/".to_string(),
        |path| path.to_string_lossy().into_owned(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_label_prefixes_the_slug_so_pi_and_omp_stay_apart() {
        assert_eq!(
            slug_for("/Users/me/app", "pi", "/cwd"),
            "pi-Users-me-app",
            "the slug is the claude one with the source label glued on"
        );
        assert_eq!(slug_for("/Users/me/app", "omp", "/cwd"), "omp-Users-me-app");
        // Underscores collapse the same way they do everywhere else.
        assert_eq!(slug_for("/a/my_app/", "pi", "/cwd"), "pi-a-my-app");
    }

    #[test]
    fn both_default_roots_are_watched_even_when_absent() {
        let adapter = PiAdapter::new();
        let watched = adapter.watch_paths();
        assert_eq!(watched.len(), 2);
        assert!(watched[0].ends_with(".pi/agent/sessions"), "{watched:?}");
        assert!(watched[1].ends_with(".omp/agent/sessions"), "{watched:?}");
    }

    #[test]
    fn an_absent_root_enumerates_empty_rather_than_failing() {
        let adapter = PiAdapter::with_roots(vec![(
            PathBuf::from("/nonexistent/stax/pi"),
            "pi".to_string(),
        )]);
        assert!(adapter.enumerate().is_empty());
    }
}
